crates/pixi_core/src/global/project/mod.rs

```diff
@@ -1304,6 +1350,78 @@ impl Project {
                 .finish())
         })
     }
+
+    /// Infer the package name from a PixiSpec (path or git) by examining build outputs
+    pub async fn infer_package_name_from_spec(
+        &self,
+        pixi_spec: &pixi_spec::PixiSpec,
+    ) -> Result<PackageName, InferPackageNameError> {
+        // Convert PixiSpec to SourceSpec and resolve it to PinnedSourceSpec
+        let pinned_source_spec = match pixi_spec {
+            pixi_spec::PixiSpec::Path(path_spec) => {
+                PinnedSourceSpec::Path(pixi_record::PinnedPathSpec {
+                    path: path_spec.path.clone(),
+                })
+            }
+            pixi_spec::PixiSpec::Git(git_spec) => {
+                // Convert GitSpec to SourceSpec
+                let source_spec = pixi_spec::SourceSpec {
+                    location: pixi_spec::SourceLocationSpec::Git(git_spec.clone()),
+                };
+
+                // Use the command dispatcher to resolve and checkout the git repository
+                let command_dispatcher = self.command_dispatcher()?;
+                let checkout = command_dispatcher
+                    .pin_and_checkout(source_spec)
+                    .await
+                    .map_err(|e| InferPackageNameError::BuildBackendMetadata(Box::new(e)))?;
+
+                checkout.pinned
+            }
+            _ => {
+                return Err(InferPackageNameError::UnsupportedSpecType);
+            }
+        };
```

Comment:
`PixiSpec` is either a binary or source package. You should call `PixiSpec::into_source_or_binary` to convert it into a source or binary spec. If it is a source spec you can call command_dispatcher.pin_and_checkout on it (regardless of the type). 

For binary specs you can have `url` or `path`. In those cases you can take the filename of the url or path and parse that as an `ArchiveIdentifier` which will also give you a package name!

Reply:
Good catch! Makes sense. A binary spec with a path is that weird match spec syntax?

Reply:
Not sure if we support it but something along these lines would be a binary path:

```
--path my-package-0-build.conda
```